Llvm 16 compatibility.

noahmartinwilliams opened the pull request:
Add support for llvm-16.

apaszke committed:
Merge branch 'llvm-12' into llvm-15

sighingnow committed:
Build with GHC 9.4.1 on llvm-15 branch, and adapt to LLVM-15.

Signed-off-by: Tao He <sighingnow@gmail.com>

sighingnow committed:
Build with GHC 9.4.1 on llvm-15 branch, and adapt to LLVM-15.

Signed-off-by: Tao He <sighingnow@gmail.com>

christianlavoie committed:
Simple fixes for LTS-22.11 and LLVM-15.0.6

tmcdonell committed:
Merge pull request #411 from sighingnow/llvm-15

tmcdonell committed:
Merge pull request #415 from christianlavoie/llvm-15

# Conflicts:
#	llvm-hs-pure/llvm-hs-pure.cabal
#	llvm-hs-pure/src/LLVM/IRBuilder/Module.hs
#	llvm-hs-pure/src/LLVM/Triple.hs
#	llvm-hs/src/LLVM/Internal/FFI/OrcJITC.cpp

tmcdonell committed:
warning police

tmcdonell committed:
add github actions CI

tmcdonell committed:
add windows CI

tmcdonell committed:
update windows ci

tmcdonell committed:
build fix for ghc-9.6

tmcdonell committed:
drop upper bound on transformers dep

tmcdonell committed:
drop shared-llvm flag

We need to link against the shared library when loading in GHCi and against the static libraries otherwise

tmcdonell committed:
Fix error 'Expect<T> must be checked before access or destruction'

tmcdonell committed:
update CI environment deps

tmcdonell committed:
update CI environment deps

noahmartinwilliams committed:
Upgrade To LLVM 16

Library has now been tested with LLVM 16.0 and can safely be used with it.

noahmartinwilliams committed:
Fix compiler errors complaining about unzip ambiguity and increment version

The library wouldn't compile because of ambiguous occurrences of unzip.

noahmartinwilliams committed:
Merge branch 'llvm-head-unstable' into llvm-16

noahmartinwilliams committed:
Increase llvm version number recommended in README.md to version 16

cassandracomar commented:
this doesn't seem to build for me due to function attributes changing. `ArgMemOnly` and friends have been moved such that they're no longer [direct function attributes and instead live under a `memory` attribute](https://releases.llvm.org/16.0.0/docs/LangRef.html#function-attributes)

cassandracomar commented:
here's a patch that fixes the build. tests are still broken because llvm-16 enforces a strict 8-bit layout width for i8 and our tests assume i8 can be laid out inside any wider type. I don't super understand the tests so it might take me awhile to get those functional.

I also haven't actually bound the memory effects related functions/types. I'm not entirely sure if it's supposed to be a separate struct or if we're supposed to reuse `ParameterAttribute`.

happy to open this as a PR to your repo.

EDIT: managed to get the memory attributes encoded/decoded into an AST by referencing `llvm/Support/ModRef.h` -- the memory attr is a couple of enums that track location and access level and get unioned/intersected to combine them. so I represented that on the HS side with an inductive data type and added folding/unfolding to the attribute encoder.

I also fixed some tests, though struggling to debug further because they die and get marked as failed without an actual error message now. 

EDIT 2: OHHH I see what's happened. I think this is a bug already existing in the package. `DICompileUnit` had 2 fields added to it at some point prior to 16.0 -- this was partially fixed by adding those [two new parameters to the C function for creating a `DICompileUnit`](https://github.com/llvm-hs/llvm-hs/blob/llvm-15/llvm-hs/src/LLVM/Internal/FFI/MetadataC.cpp#L688) but this was never updated [on the Haskell side](https://github.com/llvm-hs/llvm-hs/blob/llvm-15/llvm-hs/src/LLVM/Internal/FFI/Metadata.hs#L474). so LLVM was triggering a SIGBUS because it tried to read uninitialized memory. I updated my patch to pass those values through and encode/decode those fields.

all tests now pass after I fixed the alignment width for `i8`s and the two other problems mentioned here.

[llvm-hs-fix.patch](https://github.com/user-attachments/files/17091820/llvm-hs-fix.patch)

noahmartinwilliams commented:
Can you do a pull request on my version with this patch? It mostly works (except you need to add an ifdef to importing Distribution.Simple.PreProcess.Types for versions of cabal less than 3.12).

cassandracomar commented:
ahh good catch. yep, I'll send you a PR.

adinapoli commented:
@noahmartinwilliams @cassandracomar I just wanted to chime in to thank you both for your time and the effort you are putting in doing what you are doing. I think that having a version of `llvm-hs` (and Accelerate) working against recent versions of LLVM is very important, and so far for some other projects we have been massively slowed down / hindered by the fact we had to depend on older versions of llvm-hs (with all the chaos it entails having to also make sure the underlying toolchain is version-aligned, and nix can help, but only up to a point).

Thanks ❤️

noahmartinwilliams commented:
@adinapoli You're welcome. 

Edit: Also @cassandracomar , I added a commit to my version that you may want to check out. It won't compile without g++ version 17.

noahmartinwilliams committed:
Set g++ version to g++ 17

GunpowderGuy commented:
I get a couple of errors when updating my project from the master branch to this one. The first one may be a problem on my end, but i am positive the following one is not, as it also happens when compiling this branch by its lonesome

[error.txt](https://github.com/user-attachments/files/17534840/error.txt)

It seems like this llvm-hs is not recognizing llvm-config installed in my system. It may be because i use opensuse ( tumblewwed ) and the code is only configured for other distros such as ubuntu

noahmartinwilliams commented:
@GunpowderGuy, I've got llvm version 16 installed from source on my machine (I use Arch btw).

GunpowderGuy commented:
I got llvm version installed from the default opensuse tumbleweed repo  ( the strikethrough isnt supposed to be there ) :

diego@localhost:~>  llvm-config --version
16.0.6
diego@localhost:~>  zypper info --type package llvm16-devel
Loading repository data...
Reading installed packages...


Information for package llvm16-devel:
-------------------------------------
Repository     : Main Repository (OSS)
Name           : llvm16-devel
Version        : 16.0.6-9.1
Arch           : x86_64
Vendor         : openSUSE
Installed Size : 23.4 MiB
Installed      : Yes
Status         : up-to-date
Source package : llvm16-16.0.6-9.1.src
Upstream URL   : https://www.llvm.org/
Summary        : Header Files for LLVM
Description    : 
    This package contains library and header files needed to develop
    new native programs that use the LLVM infrastructure.

diego@localhost:~>

GunpowderGuy commented:
@noahmartinwilliams An user in the haskell matrix server pointed out the issue likely is this branch removes dynamic linking support

eg: removes the shared-llvm flag in the cabal file

Many distros have limited support for static libraries. You may not be experiencing this problem because you installed from source, but it would be nice to not requiere that

noahmartinwilliams commented:
@GunpowderGuy , You're right. I forgot about that. I can't seem to find where that flag is in either of the cabal files for llvm-hs though.

GunpowderGuy commented:
@noahmartinwilliams gekosaur over at haskell matrix shared me this : 

https://github.com/llvm-hs/llvm-hs/pull/430/files#diff-b3d9179aa3e178688f026cfd0bdefa247b227955c372f8dd4500d394ad559fe6L52-L55

https://github.com/llvm-hs/llvm-hs/pull/430/files#diff-893c0a5fbeabeeb9bc58bf002e9743c8b9b879a882171f674864daaa4dcc4fa5R152-R153

noahmartinwilliams commented:
@cassandracomar , I hate to seem pushy, but do you have that pull request ready? I don't want to do a git commit with your patch since that would mean taking credit, but I also want to get that in my repository.

GunpowderGuy commented:
@noahmartinwilliams Please, could you share me your way of installing llvm from source? It seems other haskell llvm libraries have also lost the ability to rely on dynamic linking

noahmartinwilliams commented:
@GunpowderGuy , Here's the command I used. 

`git clone --recursive https://github.com/llvm/llvm-project && cd llvm-project && git checkout llvmorg-16.0.0 && mkdir build && cd build && cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr/local -DLLVM_PARALLEL_LINK_JOBS=$(nproc) -DLLVM_BUILD_LLVM_DYLIB=True -DLLVM_LINK_LLVM_DYLIB=True ../llvm && cmake --build . -j$(nproc) && cmake --build . --target install -j$(nproc)`

GunpowderGuy commented:
@noahmartinwilliams 

Thanks for the command. It works on my end, but i replaced git checkout llvmorg-16.0.0  with git checkout llvmorg-16.0.6

After compiling llvm from source, this branch no longer complains about lack of libraries, but fails with the following error

[llvm-hs-16-error.txt](https://github.com/user-attachments/files/17611543/llvm-hs-16-error.txt)

noahmartinwilliams commented:
@GunpowderGuy Have you tried applying the patch @cassandracomar made?

GunpowderGuy commented:
> @GunpowderGuy Have you tried applying the patch @cassandracomar made?

the old error is gone, but i get this one 

/Documents/other llvm/llvm-hs> stack build
llvm-hs> configure (lib)
llvm-hs> [1 of 2] Compiling Main             ( /home/diego/Documents/other llvm/llvm-hs/llvm-hs/Setup.hs, /home/diego/Documents/other llvm/llvm-hs/llvm-hs/.stack-work/dist/x86_64-linux-tinfo6/ghc-9.2.7/setup/Main.o )
llvm-hs> 
llvm-hs> /home/diego/Documents/other llvm/llvm-hs/llvm-hs/Setup.hs:12:1: error:
llvm-hs>     Could not find module ‘Distribution.Simple.PreProcess.Types’
llvm-hs>     Perhaps you meant
llvm-hs>       Distribution.Simple.PreProcess.Unlit (from Cabal-3.6.3.0)
llvm-hs>       Distribution.Simple.PreProcess (from Cabal-3.6.3.0)
llvm-hs>       Distribution.Simple.Program.Types (from Cabal-3.6.3.0)
llvm-hs>     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
llvm-hs>    |
llvm-hs> 12 | import Distribution.Simple.PreProcess.Types (Suffix)
llvm-hs>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

noahmartinwilliams commented:
@GunpowderGuy Oh, right. I forgot that my copy has a modified version of @cassandracomar's patch. I added some lines for dealing with different versions of Cabal. 
[my-own-clearly-superior-patch.txt](https://github.com/user-attachments/files/17671017/my-own-clearly-superior-patch.txt)
Edit:
@cassandracomar , would it be alright with you if I just did a git commit with your patch and changed the email and name on the commit to yours? I need to get going on this.

zlonast commented:
I'm not sure why this pull request isn't updated, but I opened a new one with the patches above
https://github.com/llvm-hs/llvm-hs/pull/436

cassandracomar commented:
@noahmartinwilliams sorry, got sick for several months and didn't get a chance to get back to this. it's totally ok with me.

Commit:
Make llvm-hs and llvm-hs-pure Compile Again

The project should now be able to compile with new headers that llvm 16
shipped with.

cassandracomar commented:
for anyone who's interested, [I got the project to build/pass tests via nix](https://github.com/cassandracomar/llvm-hs/tree/9092d9b800ce6693cb173aae3ee4094d0f4f65bc) ([branch](https://github.com/cassandracomar/llvm-hs/tree/llvm-16)). I've left the flake fairly minimal so it takes less time to build all the dependencies but I recommend using `config.replaceStdenv` to set the stdenv for all packages to `pkgs.llvmPackages_16.libcxxStdenv`. doing so means building a lot more packages from scratch but it handles a bunch of `unused argument` warnings that I think are intended for clang 19 (as that's what ghc is built with in current nixpkgs).

of note: I had to delete most of the hooks in `llvm-hs/Setup.hs` -- nix already provides the includes and libdirs to cabal and whatever the hsc2hs hook was doing, it makes the project completely unbuildable from within a nix environment (nix shell OR derivation build context). ghc winds up unable to link against it's own `Rts.h` header -- if you tell ghc not to delete the tmp files and run the exact same clang command it output, it succeeds. so the existing hooks are doing something screwy.

if you'd like to build llvm-hs as a dependency in your own flake, you can do so (assuming you're using [`srid/haskell-flake`](https://github.com/srid/haskell-flake)) via:

```nix
perSystem = {pkgs, ...}: let
  # we need a haskell package set using llvmPackages_16 for stdenv
  # it's a bit annoying to construct one:
  callPackage = pkgs.newScope {
    haskellLib = pkgs.haskell.lib.compose;
    overrides = pkgs.haskell.packageOverrides;
    stdenv = pkgs.llvmPackages_16.libcxxStdenv;
  };
  ghc = pkgs.haskell.compiler.ghc910;
  haskellPackages = callPackage "${inputs.nixpkgs}/pkgs/development/haskell-modules" {
    inherit ghc;
    buildHaskellPackages = haskellPackages;
    compilerConfig = callPackage "${inputs.nixpkgs}/pkgs/development/haskell-modules/configuration-ghc-9.10.x.nix" {};
  };
in {
  haskellProjects.yourProject = {
    # set the package set constructed above as `basePackages`
    # -- you can use `.extends (final: prev: {})` to modify the
    #    package set as needed.
    basePackages = haskellPackages;
    
    packages = let 
      llvm-hs = pkgs.fetchFromGitHub {
        owner = "cassandracomar";
        repo = "llvm-hs";
        rev = "9092d9b800ce6693cb173aae3ee4094d0f4f65bc";
        sha256 = "sha256-mPzcNHhdlUOPY/c5fynyZ3cDCstCMtPNbX5hLWg+rLI=";
      };
    in {
      # in addition to whatever else you've set
      llvm-hs-pure.source = "${llvm-hs}/llvm-hs-pure";
      llvm-hs.source = "${llvm-hs}/llvm-hs";
    };
    
    settings = {
      # amongst any other package settings
      llvm-hs = {self, ...}: {
        extraBuildTools = with pkgs.llvmPackages_16; [self.hsc2hs libllvm.dev];
        extraPkgconfigDepends = with pkgs; [zlib xml2];
        extraLibraries = with pkgs.llvmPackages_16; [libllvm.lib];
        extraConfigureFlags = with pkgs.llvmPackages_16; [
          "--ghc-option=-pgma=${libcxxStdenv.cc}/bin/as"
          "--ghc-option=-pgmc=${libcxxStdenv.cc}/bin/clang"
          "--ghc-option=-pgmcxx=${libcxxStdenv.cc}/bin/clang"
          "--ghc-option=-pgml=${libcxxStdenv.cc}/bin/clang"
          "--ghc-option=-pgmlas=${libcxxStdenv.cc}/bin/as"
          "--ghc-option=-pgmotool=${libllvm}/bin/llvm-otool"
          "--ghc-option=-pgminstall_name_tool=${libllvm}/bin/llvm-install-name-tool"
        ];
      };
    };
  };
};
```

I'm not opening this as an independent PR against this repo as I've made no attempt to retain compatibility with a non-nix setup and I'm out of energy to figure out how to fix the old setup hooks. getting a functional environment where this package can be built is getting harder and harder -- especially now that even llvm-16 is starting to get phased out -- so I'm just glad to have a pinned environment where I know for a fact this builds.

and in case someone hasn't already provided these, I patched `accelerate-llvm-native` to work with the updates I made to `llvm-hs` last year (to get this branch functioning/passing tests with the llvm-16 API changes):
- [accelerate-llvm-fix.patch](https://github.com/user-attachments/files/19457999/accelerate-llvm-fix.patch)

and I patched `accelerate` to build with ghc-9.10:
- [accelerate-fix.patch](https://github.com/user-attachments/files/19458001/accelerate-fix.patch)

this last one I can probably open as a PR against the main repo but the first one depends on one of these `llvm-16` PRs actually being merged here and this project looks fairly unmaintained.
